File: system-test/java/src/test/java/com/optivem/shop/systemtest/legacy/mod10/acceptance/PlaceOrderNegativeTest.java
package com.optivem.shop.systemtest.legacy.mod10.acceptance;

import com.optivem.shop.systemtest.commons.providers.EmptyArgumentsProvider;
import com.optivem.shop.systemtest.legacy.mod10.acceptance.base.BaseAcceptanceTest;
import com.optivem.shop.dsl.channel.ChannelType;
import com.optivem.testing.Channel;

import org.junit.jupiter.api.TestTemplate;
import org.junit.jupiter.params.provider.ArgumentsSource;
import org.junit.jupiter.params.provider.ValueSource;

class PlaceOrderNegativeTest extends BaseAcceptanceTest {
    @TestTemplate
    @Channel({ChannelType.UI, ChannelType.API})
    @ValueSource(strings = {"3.5", "lala", "invalid-quantity"})
    void shouldRejectOrderWithNonIntegerQuantity(String nonIntegerQuantity) {
        scenario
                .when().placeOrder()
                    .withQuantity(nonIntegerQuantity)
                .then().shouldFail()
                    .errorMessage("The request contains one or more validation errors")
                    .fieldErrorMessage("quantity", "Quantity must be an integer");
    }

    @TestTemplate
    @Channel({ChannelType.UI, ChannelType.API})
    void shouldRejectOrderForNonExistentProduct() {
        scenario
                .when().placeOrder()
                    .withSku("NON-EXISTENT-SKU-12345")
                    .withQuantity(1)
                .then().shouldFail()
                    .errorMessage("The request contains one or more validation errors")
                    .fieldErrorMessage("sku", "Product does not exist for SKU: NON-EXISTENT-SKU-12345");
    }

    @TestTemplate
    @Channel({ChannelType.UI, ChannelType.API})
    @ArgumentsSource(EmptyArgumentsProvider.class)
    void shouldRejectOrderWithEmptySku(String sku) {
        scenario
                .when().placeOrder()
                    .withSku(sku)
                    .withQuantity(1)
                .then().shouldFail()
                    .errorMessage("The request contains one or more validation errors")
                    .fieldErrorMessage("sku", "SKU must not be empty");
    }

    @TestTemplate
    @Channel({ChannelType.UI, ChannelType.API})
    @ValueSource(strings = {"-10", "-1", "0"})
    void shouldRejectOrderWithNonPositiveQuantity(String quantity) {
        scenario
                .when().placeOrder()
                    .withQuantity(quantity)
                .then().shouldFail()
                    .errorMessage("The request contains one or more validation errors")
                    .fieldErrorMessage("quantity", "Quantity must be positive");
    }

    @TestTemplate
    @Channel({ChannelType.UI, ChannelType.API})
    @ArgumentsSource(EmptyArgumentsProvider.class)
    void shouldRejectOrderWithEmptyQuantity(String quantity) {
        scenario
                .when().placeOrder()
                    .withQuantity(quantity)
                .then().shouldFail()
                    .errorMessage("The request contains one or more validation errors")
                    .fieldErrorMessage("quantity", "Quantity must not be empty");
    }

    @TestTemplate
    @Channel({ChannelType.API})
    void shouldRejectOrderWithNullQuantity() {
        scenario
                .when().placeOrder()
                    .withQuantity(null)
                .then().shouldFail()
                    .errorMessage("The request contains one or more validation errors")
                    .fieldErrorMessage("quantity", "Quantity must not be empty");
    }
}

File: system-test/dotnet/SystemTests/Legacy/Mod10/AcceptanceTests/PlaceOrderNegativeTest.cs
using SystemTests.Commons.Providers;
using SystemTests.Legacy.Mod10.AcceptanceTests.Base;
using Dsl.Core.Shop;
using Optivem.Testing;
using Xunit;

namespace SystemTests.Legacy.Mod10.AcceptanceTests;

public class PlaceOrderNegativeTest : BaseAcceptanceTest
{
    [Theory]
    [ChannelData(ChannelType.API, AlsoForFirstRow = new[] { ChannelType.UI })]
    [ChannelInlineData("3.5")]
    [ChannelInlineData("lala")]
    [ChannelInlineData("invalid-quantity")]
    public async Task ShouldRejectOrderWithNonIntegerQuantity(Channel channel, string nonIntegerQuantity)
    {
        await Scenario(channel)
            .When().PlaceOrder()
                .WithQuantity(nonIntegerQuantity)
            .Then().ShouldFail()
                .ErrorMessage("The request contains one or more validation errors")
                .FieldErrorMessage("quantity", "Quantity must be an integer");
    }

    [Theory]
    [ChannelData(ChannelType.UI, ChannelType.API)]
    public async Task ShouldRejectOrderForNonExistentProduct(Channel channel)
    {
        await Scenario(channel)
            .When().PlaceOrder()
                .WithSku("NON-EXISTENT-SKU-12345")
                .WithQuantity(1)
            .Then().ShouldFail()
                .ErrorMessage("The request contains one or more validation errors")
                .FieldErrorMessage("sku", "Product does not exist for SKU: NON-EXISTENT-SKU-12345");
    }

    [Theory]
    [ChannelData(ChannelType.API, AlsoForFirstRow = new[] { ChannelType.UI })]
    [ChannelClassData(typeof(EmptyArgumentsProvider))]
    public async Task ShouldRejectOrderWithEmptySku(Channel channel, string sku)
    {
        await Scenario(channel)
            .When().PlaceOrder()
                .WithSku(sku)
                .WithQuantity(1)
            .Then().ShouldFail()
                .ErrorMessage("The request contains one or more validation errors")
                .FieldErrorMessage("sku", "SKU must not be empty");
    }

    [Theory]
    [ChannelData(ChannelType.API, AlsoForFirstRow = new[] { ChannelType.UI })]
    [ChannelClassData(typeof(EmptyArgumentsProvider))]
    public async Task ShouldRejectOrderWithEmptyQuantity(Channel channel, string quantity)
    {
        await Scenario(channel)
            .When().PlaceOrder()
                .WithQuantity(quantity)
            .Then().ShouldFail()
                .ErrorMessage("The request contains one or more validation errors")
                .FieldErrorMessage("quantity", "Quantity must not be empty");
    }

    [Theory]
    [ChannelData(ChannelType.API)]
    public async Task ShouldRejectOrderWithNullQuantity(Channel channel)
    {
        await Scenario(channel)
            .When().PlaceOrder()
                .WithQuantity(null)
            .Then().ShouldFail()
                .ErrorMessage("The request contains one or more validation errors")
                .FieldErrorMessage("quantity", "Quantity must not be empty");
    }
}
